### ISSUE TITLE
Allow rendering custom errors

### ISSUE THREAD

**Issue by jvenezia** [doorkeeper-gem/doorkeeper#844]
Hi, I need tu customise the error response with jbuilder. The only way to do that is with `doorkeeper_unauthorized_render_options`:

``` ruby
    def doorkeeper_unauthorized_render_options(error: nil)
      @error_key = :unauthorized
      @error_message = 'The access token is not valid.'
      {json: render_to_string('api/open/v1/errors/general.json.jbuilder')}
    end
```

I would prefer to render the error my self to be consistant with the rest of my app:

``` ruby
    def doorkeeper_unauthorized_render_options(error: nil)
      render_general_error(:unauthorized, 'The access token is not valid.', 401)
    end

    def render_general_error(key, message, status)
      @error_key = key
      @error_message = message
      render 'api/open/v1/errors/general.json.jbuilder', status: status
    end
```

But doing so legitimately raises an `AbstractController::DoubleRenderError`.

Is there already a way to do that?

An option to raise an `Doorkeeper::Unauthorized` exception that I handle my self would be a nice solution.

**Comment by dino115** [comment]
+1

**Comment by wRadion** [comment]
:+1:

Especially for the `unauthorized` error you presented; because it is the only one not specified by the OAuth2 specifications, so we might be wanting to handle it ourselves.

**Comment by nbulaj** [comment]
Hi @ jvenezia. Why you need to customize your error response with the jBuilder view ? Why don't just return a hash (JSON) ?

The error occurs because Doorkeeper already called `render` and passed the result of the `doorkeeper_unauthorized_render_options` method (take a look at https://github.com/doorkeeper-gem/doorkeeper/blob/5ea4575863781940fe6996692c6b41f2ad71d90e/lib/doorkeeper/rails/helpers.rb#L32). So you can try to return some options for render in your `render_general_error` method, like:

```ruby
def render_general_error(key, message, status)
      @error_key = key
      @error_message = message
      layout: 'api/open/v1/errors/general.json.jbuilder', error_key: key # or try with :template
end
```
I hope you caught the direction of my thoughts.

**Comment by jvenezia** [comment]
Hi @nbulaj 

My point is that I want to handle this my self to be consistant to the rest of my app.

Raising an exception like `Doorkeeper::Unauthorized` will allow me to handle this error the same way that I render my other errors.

**Comment by nbulaj** [comment]
@jvenezia ok, I understand. Did you try my advise with the render options?

**Comment by jvenezia** [comment]
Actually It's working with `doorkeeper_unauthorized_render_options` (My first example).

I think your solution will work as well, but I don't want to change my code to fit with doorkeeper.

I am not trying to make this working but I am suggesting a new feature.

**Comment by tute** [comment]
> Actually It's working with `doorkeeper_unauthorized_render_options` (My first example).

Does it fully work, or does it raise `AbstractController::DoubleRenderError`? If it works, do you think we can still make improvements in doorkeeper to accommodate better these overrides? Thanks!

**Comment by nbulaj** [comment]
> Does it fully work, or does it raise AbstractController::DoubleRenderError?

@tute - it return an Exception because @jvenezia doesn't want to change the implementation of the `render_general_error` method and wants to use it in the Doorkeeper `doorkeeper_unauthorized_render_options` method (as he wrote in the first message):

```ruby
def doorkeeper_unauthorized_render_options(error: nil)
  render_general_error(:unauthorized, 'The access token is not valid.', 401)
end

def render_general_error(key, message, status)
  @error_key = key
  @error_message = message
  render 'api/open/v1/errors/general.json.jbuilder', status: status
end
```

And Doorkeeper calls `render` with `doorkeeper_unauthorized_render_options`, that also calls `render` - so it produce the `AbstractController::DoubleRenderError` exception.

The way it may be improved is only to change Doorkeeper `doorkeeper_render_error_with` method to be ovverridable I think.

**Comment by aalvarado** [comment]
I think an option to do this could be:

```ruby
class BaseController < ActionController::API
  class InvalidTokenError < StandardError
    def initialize(msg="some message")
      super(msg)
    end
  end

  rescue_from(InvalidTokenError) do |e|
    render json: { error: e.message }, status: 401
  end


  def doorkeeper_render_error_with(error)
    if error.is_a? Doorkeeper::OAuth::InvalidTokenResponse
      raise InvalidTokenError
    else
      super
    end
  end
end
```

this only takes care of one error, if you want to handle more errors I would probably use a class that can map one of the Doorkeeper error classes that don't inherit from `StandardError` and raise a real exception. 

something like:

```ruby
class DoorkeeperExceptionRaiser
  HANDLED_ERRORS = {
    Doorkeeper::OAuth::InvalidTokenResponse => InvalidTokenError,
    Doorkeeper::OAuth::ForbiddenTokenResponse => ForbiddenTokenError,
    # ... More mappings.
  }
  
  def initialize(doorkeeper_error)
    @doorkeeper_error = doorkeeper_error
  end

  def raise_if_handled
    if doorkeeper_error_classes.include? doorkeeper_error.class
      raise HANDLED_ERRORS[doorkeeper_error.class]
    end
  end

  def doorkeeper_error_classes
    HANDLED_ERRORS.keys
  end
end
```

and then use it on `doorkeeper_render_error_with` and call super afterwards.


```ruby
def doorkeeper_render_error_with(error)
  DoorkeeperExceptionRaiser.new(error).raise_if_handled
  super
end
```

**Comment by nbulaj** [comment]
Implemented in #1140 and would be released in 5.x.x or 5.0.x release :tada:

**Comment by aalvarado** [comment]
Awesome, thanks @nbulaj
